File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/SecurityInvariantPlanRequirements.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.security.*;
import java.util.*;

/**
 * Derives the minimum plan-level security contract from the authorized plan
 * shape.
 */
public final class SecurityInvariantPlanRequirements {
	private SecurityInvariantPlanRequirements() {
	}

	public static SemanticPlan attach(SemanticPlan plan) {
		return attach(plan, null);
	}

	public static SemanticPlan attach(SemanticPlan plan, Iterable<String> capabilityInvariants) {
		Objects.requireNonNull(plan);
		var ids = new HashSet<>(plan.effectiveSecurityInvariants());
		if (capabilityInvariants != null)
			for (var id : capabilityInvariants) {
				if (!SecurityInvariantRegistry.contains(id))
					throw new IllegalStateException("Unknown capability security invariant '" + id + "'.");
				ids.add(id);
			}
		collect(plan.root(), ids);
		if (ids.isEmpty())
			throw new IllegalStateException("A semantic plan cannot become executable without a security contract.");
		return new SemanticPlan(plan.root(), ids.stream().sorted().toList(), plan.authorizationBinding());
	}

	private static void collect(SemanticPlanNode node, Set<String> ids) {
		ids.add(SecurityInvariantIds.AUTHORIZATION_REQUIRED);
		ids.add(SecurityInvariantIds.PARAMETERIZED_VALUES);
		ids.add(SecurityInvariantIds.PLAN_CACHE_CONTEXT_ISOLATION);
		if (!node.fields().isEmpty())
			ids.add(SecurityInvariantIds.FIELD_VISIBILITY);
		if (!node.children().isEmpty())
			ids.add(SecurityInvariantIds.RELATIONSHIP_VISIBILITY);
		if (node.authorization() != null)
			ids.add(SecurityInvariantIds.RUNTIME_AUTHORIZATION);
		node.children().forEach(c -> collect(c, ids));
	}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/SecurityInvariantPlanRequirements.cs
using Foundgine.Core.Semantic.Security;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Derives the minimum plan-level security contract from the authorized plan
/// shape. Capability-specific requirements can be supplied explicitly and are
/// preserved rather than replaced.
/// </summary>
public static class SecurityInvariantPlanRequirements
{
    public static SemanticPlan Attach(SemanticPlan plan) => Attach(plan, null);

    /// <summary>
    /// Attaches the plan obligations plus the invariants declared by the
    /// capability/security contract. The supplied capability invariants are
    /// additive: plan shape may require more guarantees, never fewer.
    /// </summary>
    public static SemanticPlan Attach(
        SemanticPlan plan,
        IEnumerable<string>? capabilityInvariants)
    {
        ArgumentNullException.ThrowIfNull(plan);

        var ids = new HashSet<string>(
            plan.RequiredSecurityInvariants ?? [],
            StringComparer.Ordinal);

        if (capabilityInvariants is not null)
        {
            foreach (var id in capabilityInvariants)
            {
                if (!SecurityInvariantRegistry.Contains(id))
                    throw new InvalidOperationException(
                        $"Unknown capability security invariant '{id}'.");
                ids.Add(id);
            }
        }

        Collect(plan.Root, ids);

        if (ids.Count == 0)
            throw new InvalidOperationException(
                "A semantic plan cannot become executable without a security contract.");

        return plan with
        {
            RequiredSecurityInvariants = ids
                .OrderBy(x => x, StringComparer.Ordinal)
                .ToArray()
        };
    }

    private static void Collect(SemanticPlanNode node, ISet<string> ids)
    {
        ids.Add(SecurityInvariantIds.AuthorizationRequired);
        ids.Add(SecurityInvariantIds.ParameterizedValues);
        ids.Add(SecurityInvariantIds.PlanCacheContextIsolation);

        if (node.Fields.Count > 0)
            ids.Add(SecurityInvariantIds.FieldVisibility);

        if (node.Children.Count > 0)
            ids.Add(SecurityInvariantIds.RelationshipVisibility);

        if (node.Authorization is not null)
        {
            ids.Add(SecurityInvariantIds.RuntimeAuthorization);
        }

        foreach (var child in node.Children)
            Collect(child, ids);
    }
}
